### PR TITLE
benchmark for groupby operation

### PR DESCRIPTION
**Pull request by Jarvi-Izana**
#2210 pr is too large and it contains too many topics and therefore it's hard to be reviewed.

I separate benchmark test code from #2210 in order to polish it first and then share it among other branches. 

Let's make it into master first, and then I can start bench testing other ideas of optimization by branching from master to share the same benchmark test code.

### CODE REVIEW

**Comment by Jarvi-Izana** [comment]
If we have more benchmark code, then we need to move the `benchmark` dir under top-level dir. Currently, I don't think we need.

File: test/benchmark/sframe_bench_aggregate.cpp [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+#include <thread>
+#include <random>
+
+namespace turi {
+
+using row_gen_func_t = std::function<std::vector<flexible_type>(size_t)>;
+
+/* generate sframe */
+/* I need to make this a fixture */
+static sframe bench_test_sframe_generator(
```
Comment (hoytak) [inline test/benchmark/sframe_bench_aggregate.cpp]: Use random sframe generator code in <core/storage/sframe_data/testing_utils.hpp> -- that would be easier to use. We don't want to duplicate code if we have good reason not to.

**Comment by hoytak** [comment]
I think this is ready -- want to merge it?

**Comment by Jarvi-Izana** [comment]
I still have some changes. Let me clean them up and then we are good to go!

**Comment by Jarvi-Izana** [comment]
Refactored my code and it should be very easy to add more benchmark test cases.

sample output. A force update is due to I mistakenly amended all my changes to a previous commit.
```
=========== count on binary categorical data =============

nrows: 100000
reps: 3
avg time to run w/  2 threads: 42.1337 ms.
avg time to run w/  4 threads: 30.247 ms.
avg time to run w/  8 threads: 28.779 ms.
avg time to run w/ 12 threads: 26.7163 ms.
========================== END ===========================

=========== bench_test_aggreate_min summary ============

nrows: 100000
reps: 3
users: 1000
avg time to run w/  2 threads: 187.639 ms.
avg time to run w/  4 threads: 163.525 ms.
avg time to run w/  8 threads: 154.321 ms.
avg time to run w/ 12 threads: 145.579 ms.
========================== END ===========================

=========== bench_test_aggreate_avg summary ============

nrows: 100000
reps: 3
users: 1000
avg time to run w/  2 threads: 180.494 ms.
avg time to run w/  4 threads: 158.512 ms.
avg time to run w/  8 threads: 156.701 ms.
avg time to run w/ 12 threads: 146.346 ms.
========================== END ===========================
```

**Comment by Jarvi-Izana** [comment]
ready to go! @hoytak

**Comment by Jarvi-Izana** [comment]
added data generator for skewed data distribution. 

It will be later used for validating the `rwlock` implementation perf since if a key appears a lot more frequent, then it will cause more contentions when sframe performs aggregations under multithreading context.

File: test/benchmark/sframe_bench_aggregate.cpp [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+ * 3. number of unique keys except count operator
+ *
+ * the generated sframe will be reused for different
+ * sized thread pool settings.
+ * */
+void bench_test_aggreate_summary_count_bin(size_t nrows, size_t reps) {
+  ASSERT_MSG(reps > 0, "reps shouldn't be 0");
+  ASSERT_MSG(nrows > 0, "nrows shouldn't be 0");
+
+  sframe sf = make_random_sframe(nrows, "b", false);
```
Comment (Jarvi-Izana) [inline test/benchmark/sframe_bench_aggregate.cpp]: adopted Hoyt's advice.

File: test/benchmark/sframe_bench_aggregate.cpp [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+        ret.emplace_back(values[itr - samples.begin()]);
+      } else {
+        ret.emplace_back(dist_dat(gen));
+      }
+    }
+
+    return ret;
+  };
+
+  sframe sf = make_testing_sframe(keys, {flexible_type(T()).get_type()}, nrows,
```
Comment (Jarvi-Izana) [inline test/benchmark/sframe_bench_aggregate.cpp]: added a new utility routine into `testing_utils`.

**Comment by hoytak** [comment]
Looks good!  Go ahead and merge it!

**Pull request merged**